crates/flux-refineck/src/checker.rs

```diff
@@ -229,20 +231,70 @@ impl<'ck, 'genv, 'tcx> Checker<'ck, 'genv, 'tcx, RefineMode> {
 /// `SubFn` lets us reuse _most_ of the same code for `check_fn_subtyping` for both the case where
 /// we have an early-bound function signature (e.g., for a trait method???) and versions without,
 /// e.g. a plain closure against its FnTraitPredicate obligation.
+#[derive(Debug)]
 pub enum SubFn {
-    Poly(EarlyBinder<rty::PolyFnSig>, rty::GenericArgs),
+    Poly(DefId, EarlyBinder<rty::PolyFnSig>, rty::GenericArgs),
     Mono(rty::PolyFnSig),
 }
 
 impl SubFn {
     pub fn as_ref(&self) -> &rty::PolyFnSig {
         match self {
-            SubFn::Poly(sig, _) => sig.skip_binder_ref(),
+            SubFn::Poly(_, sig, _) => sig.skip_binder_ref(),
             SubFn::Mono(sig) => sig,
         }
     }
 }
 
+/// The function `hoist_with_binders` uses the `LocalHoister` machinery to convert
+/// a (closure) function template _without_ binders, e.g. `fn ({v.i32 | *}) -> {v.i32|*})`
+/// into one _with_ input binders, e.g. `forall <a:int>. fn ({i32[a]|*}) -> {v.i32|*}`
+/// after which the hole-filling machinery can be used to fill in the holes.
+/// This lets us get "dependent signatures" where the output can refer to the input.
+/// e.g. see `tests/pos/surface/closure09.rs`
+fn hoist_input_binders(poly_sig: &PolyFnSig) -> PolyFnSig {
+    // println!("TRACE: hoist_input_binders (0): {poly_sig:?}");
+    let original_vars = poly_sig.vars().to_vec();
+    let fn_sig = poly_sig.skip_binder_ref();
+    let mut delegate = LocalHoister::new(original_vars);
+    let mut hoister = Hoister::with_delegate(&mut delegate).transparent();
+
+    let inputs = fn_sig
+        .inputs()
+        .iter()
+        .map(|ty| hoister.hoist(ty))
+        .collect_vec();
+
+    delegate.bind(|_vars, preds| {
+        let (mut holes, mut preds): (Vec<_>, Vec<_>) = preds
+            .into_iter()
+            .partition(|pred| matches!(pred.kind(), ExprKind::Hole(HoleKind::Pred)));
+        if let Some(hole) = holes.pop() {
+            preds.push(hole);
+        }
```

Comment:
maybe
```suggestion
    delegate.bind(|_vars, mut preds| {
        let mut seen_hole = false;
        preds.retain(|pred| {
          if let ExprKind::Hole(HoleKind::Pred) = pred.kind() {
              if seen_hole {
                  false
              } else {
                  seen_hole = true;
                  true
              }
          } else {
              true
          }
        });
            
```

It's technically more efficient because it doesn't have to allocate extra vectors, but I dunno, looks kind of ugly.

Reply:
This one is cute.

```rust
    delegate.bind(|_vars, mut preds| {
        let mut keep_hole = true;
        preds.retain(|pred| {
          if let ExprKind::Hole(HoleKind::Pred) = pred.kind() {
              std::mem::replace(&mut keep_hole, false)
          } else {
              true
          }
        });
```